File: packages/prebid_mobile_sdk_gam/android/src/main/kotlin/com/prebid/prebid_mobile_sdk_gam/GamInterstitialManager.kt
package com.prebid.prebid_mobile_sdk_gam

import android.app.Activity
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.prebid.mobile.api.exceptions.AdException
import org.prebid.mobile.api.rendering.InterstitialAdUnit
import org.prebid.mobile.api.rendering.listeners.InterstitialAdUnitListener
import org.prebid.mobile.eventhandlers.GamInterstitialEventHandler

/// Handles GAM-rendered interstitials over the `prebid_mobile_sdk_gam/interstitial`
/// method channel. Each ad is keyed by an `adId` allocated on the Dart side;
/// native events are pushed back over the same channel.
class GamInterstitialManager(
    messenger: BinaryMessenger,
    private val activityProvider: () -> Activity?,
) : MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, "prebid_mobile_sdk_gam/interstitial")
    private val ads = mutableMapOf<Long, InterstitialAdUnit>()

    init {
        channel.setMethodCallHandler(this)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        ads.values.forEach { it.destroy() }
        ads.clear()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val adId = (args?.get("adId") as? Number)?.toLong()

        when (call.method) {
            "load" -> {
                val activity = activityProvider()
                if (activity == null) {
                    result.error("no_activity", "No attached Activity to load the interstitial", null)
                    return
                }
                if (adId == null) {
                    result.error("no_ad_id", "Missing adId", null)
                    return
                }
                val configId = args?.get("configId") as? String ?: ""
                val gamAdUnitId = args?.get("gamAdUnitId") as? String ?: ""

                val eventHandler = GamInterstitialEventHandler(activity, gamAdUnitId)
                val adUnit = InterstitialAdUnit(activity, configId, eventHandler)

                adUnit.setInterstitialAdUnitListener(object : InterstitialAdUnitListener {
                    override fun onAdLoaded(unit: InterstitialAdUnit) = send(adId, "onAdLoaded")
                    override fun onAdFailed(unit: InterstitialAdUnit, e: AdException?) =
                        send(adId, "onAdFailed", e?.message ?: "Unknown error")
                    override fun onAdDisplayed(unit: InterstitialAdUnit) = send(adId, "onAdDisplayed")
                    override fun onAdClosed(unit: InterstitialAdUnit) = send(adId, "onAdClosed")
                    override fun onAdClicked(unit: InterstitialAdUnit) = send(adId, "onAdClicked")
                })

                ads[adId] = adUnit
                adUnit.loadAd()
                result.success(null)
            }

            "show" -> {
                ads[adId]?.show()
                result.success(null)
            }

            "destroy" -> {
                ads.remove(adId)?.destroy()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun send(adId: Long, event: String, error: String? = null) {
        val payload = mutableMapOf<String, Any?>("adId" to adId)
        if (error != null) payload["error"] = error
        channel.invokeMethod(event, payload)
    }
}

File: packages/prebid_mobile_sdk_gam/ios/prebid_mobile_sdk_gam/Sources/prebid_mobile_sdk_gam/GamInterstitialManager.swift
import Flutter
import UIKit
import PrebidMobile
import PrebidMobileGAMEventHandlers

/// Handles GAM-rendered interstitials over the `prebid_mobile_sdk_gam/interstitial`
/// method channel. Each ad is keyed by an `adId` allocated on the Dart side;
/// native events are pushed back over the same channel.
class GamInterstitialManager: NSObject, InterstitialAdUnitDelegate {

    private let channel: FlutterMethodChannel
    private var ads: [Int: InterstitialRenderingAdUnit] = [:]
    private var adIdByUnit: [ObjectIdentifier: Int] = [:]

    init(messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: "prebid_mobile_sdk_gam/interstitial",
            binaryMessenger: messenger
        )
        super.init()
        channel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result)
        }
    }

    private func handle(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let adId = args?["adId"] as? Int

        switch call.method {
        case "load":
            guard let adId = adId else {
                result(FlutterError(code: "no_ad_id", message: "Missing adId", details: nil))
                return
            }
            let configId = args?["configId"] as? String ?? ""
            let gamAdUnitId = args?["gamAdUnitId"] as? String ?? ""

            let eventHandler = GAMInterstitialEventHandler(adUnitID: gamAdUnitId)
            let adUnit = InterstitialRenderingAdUnit(configID: configId, eventHandler: eventHandler)
            adUnit.adFormats = [.banner]
            adUnit.delegate = self

            ads[adId] = adUnit
            adIdByUnit[ObjectIdentifier(adUnit)] = adId
            adUnit.loadAd()
            result(nil)

        case "show":
            if let adId = adId,
               let adUnit = ads[adId],
               let root = UIApplication.shared.keyWindow?.rootViewController {
                adUnit.show(from: root)
            }
            result(nil)

        case "destroy":
            if let adId = adId, let adUnit = ads.removeValue(forKey: adId) {
                adIdByUnit.removeValue(forKey: ObjectIdentifier(adUnit))
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func send(_ interstitial: InterstitialRenderingAdUnit, _ event: String, error: String? = nil) {
        guard let adId = adIdByUnit[ObjectIdentifier(interstitial)] else { return }
        var payload: [String: Any] = ["adId": adId]
        if let error = error {
            payload["error"] = error
        }
        channel.invokeMethod(event, arguments: payload)
    }

    // MARK: - InterstitialAdUnitDelegate

    func interstitialDidReceiveAd(_ interstitial: InterstitialRenderingAdUnit) {
        send(interstitial, "onAdLoaded")
    }

    func interstitial(_ interstitial: InterstitialRenderingAdUnit, didFailToReceiveAdWithError error: Error?) {
        send(interstitial, "onAdFailed", error: error?.localizedDescription ?? "Unknown error")
    }

    func interstitialWillPresentAd(_ interstitial: InterstitialRenderingAdUnit) {
        send(interstitial, "onAdDisplayed")
    }

    func interstitialDidDismissAd(_ interstitial: InterstitialRenderingAdUnit) {
        send(interstitial, "onAdClosed")
    }

    func interstitialDidClickAd(_ interstitial: InterstitialRenderingAdUnit) {
        send(interstitial, "onAdClicked")
    }
}
